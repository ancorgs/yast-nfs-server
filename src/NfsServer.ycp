/**
 * File:
 *   modules/NfsServer.ycp
 *
 * Module:
 *   Configuration of nfs_server
 *
 * Summary:
 *   NFS server configuration data, I/O functions.
 *
 * Authors:
 *   Martin Vidner <mvidner@suse.cz>
 *
 * $Id$
 *
 */

{
    module "NfsServer";
    textdomain "nfs_server";

    import "Progress";
    import "Report";
    import "Service";
    import "Summary";
    import "SuSEFirewall";
    import "Wizard";


    /* default value of settings modified */
    global boolean modified = false;


    /**
     * Function sets internal variable, which indicates, that any
     * settings were modified, to "true"
     */
    global define void SetModified () {
        modified = true;
    }

    /**
     * Functions which returns if the settings were modified
     * @return boolean  settings were modified
     */
    global define boolean GetModified () {
        return modified;
    }


    global define void Set (map settings);

    /**
     * Required packages for this module to operate
     *
     */
    global list<string> required_packages = ["nfs-utils"];


    /**
     * Write only, used during autoinstallation.
     * Don't run services and SuSEconfig, it's all done at one place.
     */
    global boolean write_only = false;


    /**
     * Should the server be started?
     * New since 9.0: Exports are independent of this setting.
     */
    global boolean start = false;

    /**
     * @example
     * [
     *   $[
     *     "mountpoint": "/projects",
     *     "allowed": [ "*.local.domain(ro)", "@trusted(rw)"]
     *   ],
     *   $[ ... ],
     *   ...
     * ]
     *
     */
    global list<map<string,any> > exports = [];

    /**
     * Do we have nfslock? (nfs-utils: yes, nfs-server: no)
     */
    boolean have_nfslock = true;

    /**
     * Get all NFS server configuration from a map.
     * When called by nfs_server_auto (preparing autoinstallation data)
     * the map may be empty.
     * @param settings	$["start": "exports":]
     * @return	success
     * @see	exports
     */
    global define boolean Import (map settings) ``{
	/*
	if (size (settings) == 0)
	{
	    // Reset - just continue with Set (#24544).
	}*/

	// To avoid enabling nfslock if it does not exist during autoinstall
	have_nfslock = (boolean) SCR::Read (.init.scripts.exists, "nfslock");
	Set (settings);
	return true;
    }

    /**
     * Set the variables just as is and without complaining
     * @param  settings
     * @return void
     */
    global define void Set (map settings)
	``{
	start = lookup (settings, "start_nfsserver", false);
	exports = lookup (settings, "nfs_exports", []);
	return;
    }


    /**
     * Dump the NFS settings to a map, for autoinstallation use.
     * @return	$["start": "exports":]
     * @see	exports
     */
    global define map Export () ``{
	return $[
	    "start_nfsserver": start,
	    "nfs_exports": exports,
	    ];
    }

    /**
     * Reads NFS settings from the SCR (.etc.exports)
     * @return true on success
     */
    global define boolean Read () ``{
	start = Service::Enabled("nfsserver");
	exports = (list<map<string,any> >) SCR::Read (.etc.exports);
	have_nfslock = (boolean) SCR::Read (.init.scripts.exists, "nfslock");

	SuSEFirewall::Read ();

	return exports != nil;
    }


    /**
     * Saves /etc/exports and creates missing directories.
     * @return true on success
     */
    global define boolean WriteExports () ``{
	    // create missing directories.
	    foreach (map<string, any> entry, exports, ``{
		string directory = (string) (entry["mountpoint"]:nil);
		if (SCR::Read (.target.dir, directory) == nil)
		{
		    if (! (boolean) SCR::Execute (.target.mkdir, directory))
		    {
			// not fatal - write other dirs.
			Report::Warning (sformat(_("Unable to create a missing directory:\n%1"), directory));
		    }
		}
	    });

	    // (the backup is now done by the agent)
	    if (! SCR::Write(.etc.exports, exports))
	    {
		// error popup message
		Report::Error (_("Unable to write to /etc/exports.
No changes will be made to the
exported directories.\n"));
		return false;
	    }

	    return true;
    }

    /**
     * Saves NFS server configuration. (exports(5))
     * Creates any missing directories.
     * @return true on success
     */
    global define boolean Write () ``{
	// if there is still work to do, don't return false immediately
	// but remember the error
	boolean ok = true;

	// dialog label
	Progress::New (_("Writing NFS Server Configuration..."), " ", 2, [
	    // progress stage label
	    _("Save /etc/exports"),
	    // progress stage label
	    _("Restart services"),
	   ], [
	    // progress step label
	    _("Saving /etc/exports..."),
	    // progress step label
	    _("Restarting services..."),
	    // final progress step label
	    _("Finished") ],
	    "" );

	// help text
	if (!write_only)
		Wizard::RestoreHelp(_("Writing NFS server settings"));

	Progress::NextStage ();

	// Independent of @ref start because of Heartbeat (#27001).
	if (! WriteExports ())
	{
	    Progress::Finish ();
	    return false;
	}

	Progress::NextStage ();

	if (!start)
	{
	    if (!write_only)
		Service::Stop ("nfsserver");

	    if (! Service::Disable("nfsserver"))
	    {
		Report::Error(Service::Error());
		ok = false;
	    }
	    if (have_nfslock)
	    {
		if (!write_only)
		    Service::Stop("nfslock");
		if (! Service::Disable("nfslock"))
		{
		    Report::Error(Service::Error());
		    ok = false;
		}
	    }
	}
	else
	{
	    if (! Service::Enable("portmap"))
	    {
		Report::Error (Service::Error());
		ok = false;
	    }
	    if (have_nfslock)
	    {
		if (! Service::Enable("nfslock"))
		{
		    Report::Error(Service::Error());
		    ok = false;
		}
	    }
	    if (! Service::Enable("nfsserver"))
	    {
		Report::Error(Service::Error());
		ok = false;
	    }

	    if (!write_only)
	    {
		if (Service::Status("portmap") != 0 )
		{
		    // portmap must not be started if it is running already (see bug # 9999)
		    Service::Start("portmap");
		}

		Service::Stop("nfsserver");
		if (have_nfslock)
		{
		    Service::Restart("nfslock");
		}
		Service::Start("nfsserver");

		if (Service::Status("nfsserver") != 0)
		{
		    // error popup message
		    Report::Error (_("Unable to restart the NFS server.
Your changes will be active after reboot.
"));
		    ok = false;
		}
	    }
	}

	SuSEFirewall::WriteOnly ();
	if (!write_only)
	{
	    SuSEFirewall::ActivateConfiguration ();
	}

	Progress::NextStage ();

	return ok;
    }

    /**
     * @return A summary for autoyast
     */
    global define string Summary() ``{
	string summary = "";
	// summary header; directories exported by NFS
	summary = Summary::AddHeader (summary, _("NFS Exports"));
	if (size (exports) > 0)
	{
	    foreach(map e, exports, ``{
		summary = Summary::OpenList (summary);
		summary = Summary::AddListItem (summary, e["mountpoint"]:"");
		summary = Summary::CloseList (summary);
	    });
	}
	else
	{
	    summary = Summary::AddLine (summary, Summary::NotConfigured ());
	}
	return summary;
    }


	/**
	* Return required packages for auto-installation
	* @return map of packages to be installed and to be removed
	*/
	global define map AutoPackages() ``{
		return ($["install": required_packages, "remove": []]);
	}



  /* The end of the definitions */

}
