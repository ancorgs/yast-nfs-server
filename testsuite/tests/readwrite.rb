# encoding: utf-8

# Module:
#   NFS server configuration
#
# Summary:
#   Testsuite
#
# Authors:
#   Martin Vidner <mvidner@suse.cz>
#
# $Id$
module Yast
  class ReadwriteClient < Client
    def main
      # testedfiles: NfsServer.ycp Service.ycp Report.ycp Testsuite.ycp

      Yast.include self, "testsuite.rb"

      @I_READ = { "target" => { "size" => 0 } }
      @I_WRITE = {}
      @I_EXEC = { "target" => { "bash_output" => {} } }
      TESTSUITE_INIT([@I_READ, @I_WRITE, @I_EXEC], nil)

      Yast.import "NfsServer"
      Yast.import "Report"

      Report.DisplayErrors(false, 0)

      @service_on = { "start" => ["3", "5"], "stop" => ["3", "5"] }
      @service_off = { "start" => [], "stop" => [] }
      @exports = [
        { "allowed" => ["proj*.local.domain(rw)"], "mountpoint" => "/projects" },
        {
          "allowed"    => ["*.local.domain(ro)", "@trusted(rw)"],
          "mountpoint" => "/usr"
        },
        { "allowed" => ["(ro,insecure,all_squash)"], "mountpoint" => "/pub" }
      ]
      @READ = {
        # Runlevel:
        "init"   => {
          "scripts" => {
            "exists"   => true,
            "runlevel" => {
              "rpcbind"   => @service_on,
              "nfsserver" => @service_on
            },
            # their contents is not important for ServiceAdjust
            "comment"  => {
              "rpcbind"   => {},
              "nfsserver" => {}
            }
          }
        },
        # 	// targetpkg:
        # 	"targetpkg": $[
        # 	    // autofs
        # 	    "installed": true,
        # 	    ],
        # NfsServer itself:
        "etc"    => {
          "exports"   => @exports,
          "sysconfig" => {}
        },
        "target" => {
          "dir"  => nil,
          # pretend none exist
          "stat" => { "dummy" => true }
        }
      }

      @WRITE = {}

      @WRITE_KO = { "etc" => { "exports" => false } }

      @EXECUTE = {
        "target" => {
          "bash_output" => { "exit" => 0, "stdout" => "", "stderr" => "" },
          "mkdir"       => true
        }
      }

      NfsServer.write_only = false

      DUMP("Read")
      TEST(lambda { NfsServer.Read }, [@READ, @WRITE, @EXECUTE], nil)
      DUMP("Write OK")
      TEST(lambda { NfsServer.Write }, [@READ, @WRITE, @EXECUTE], nil)
      DUMP("Write KO")
      TEST(lambda { NfsServer.Write }, [@READ, @WRITE_KO, @EXECUTE], nil)

      nil
    end
  end
end

Yast::ReadwriteClient.new.main
