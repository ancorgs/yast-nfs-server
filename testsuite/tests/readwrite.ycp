/**
 * Module:
 *   NFS client configuration
 *
 * Summary:
 *   Testsuite
 *
 * Authors:
 *   Martin Vidner <mvidner@suse.cz>
 *
 * $Id$
 */
{
    include "testsuite.ycp";
    import "NfsServer";
    import "Report";

    Report::DisplayErrors (false, 0);

    map service_on = $[
	"start": [ "3", "5"],
	"stop":	 [ "3", "5"],
	];
    map service_off = $[
	"start": [],
	"stop":	 [],
	];
    list exports = [$["allowed":["proj*.local.domain(rw)"],
		      "mountpoint":"/projects"],
		    $["allowed":["*.local.domain(ro)", "@trusted(rw)"],
		      "mountpoint":"/usr"],
		    $["allowed":["(ro,insecure,all_squash)"],
		      "mountpoint":"/pub"]];
    map READ = $[
	// Runlevel:
	"init": $[
	    "scripts": $[
		"exists": true,
		"runlevel": $[
		    "portmap": service_on,
		    "nfsserver": service_on,
		    ],
		// their contents is not important for ServiceAdjust
		"comment": $[
		    "portmap": $[],
		    "nfsserver": $[],
		    ],
		],
	    ],
    /*
	// targetpkg:
	"targetpkg": $[
	    // autofs
	    "installed": true,
	    ],
    */
	// NfsServer itself:
	"etc": $[
	    "exports": exports,
	    "sysconfig": $[
		],
	    ],
	"target": $[
	    "dir": nil,		// pretend none exist
	    ]
	];

    map WRITE = $[
	];

    map WRITE_KO = $[
	"etc": $[
	    "exports": false,
	    ],
	];

    map EXECUTE = $[
	"target": $[
	    "bash_output": $[
		"exit": 0,
		"stdout": "",
		"stderr": "",
		],
	    "mkdir": true,
	    ],
	];

    DUMP ("Read");
    TEST (``(NfsServer::Read ()), [READ, WRITE, EXECUTE], nil);
    DUMP ("Write OK");
    TEST (``(NfsServer::Write ()), [READ, WRITE, EXECUTE], nil);
    DUMP ("Write KO");
    TEST (``(NfsServer::Write ()), [READ, WRITE_KO, EXECUTE], nil);
}
