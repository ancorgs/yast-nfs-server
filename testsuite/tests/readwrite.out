Dump	Read
Read	.init.scripts.exists "nfsserver" true
Read	.init.scripts.runlevel "nfsserver" $["nfslock":$["start":["3", "5"], "stop":["3", "5"]], "nfsserver":$["start":["3", "5"], "stop":["3", "5"]], "rpcbind":$["start":["3", "5"], "stop":["3", "5"]]]
Read	.etc.exports [$["allowed":["proj*.local.domain(rw)"], "mountpoint":"/projects"], $["allowed":["*.local.domain(ro)", "@trusted(rw)"], "mountpoint":"/usr"], $["allowed":["(ro,insecure,all_squash)"], "mountpoint":"/pub"]]
Read	.init.scripts.exists "nfslock" true
Read	.sysconfig.nfs.NFS4_SUPPORT nil
Read	.sysconfig.nfs.NFS_SECURITY_GSS nil
Read	.init.scripts.exists "SuSEfirewall2_init" true
Read	.init.scripts.runlevel "SuSEfirewall2_init" $["nfslock":$["start":["3", "5"], "stop":["3", "5"]], "nfsserver":$["start":["3", "5"], "stop":["3", "5"]], "rpcbind":$["start":["3", "5"], "stop":["3", "5"]]]
Read	.init.scripts.exists "SuSEfirewall2_init" true
Execute	.target.bash "/etc/init.d/SuSEfirewall2_init status" $["TERM":"raw"] 0
Return	true
Dump	Write OK
Read	.target.dir "/projects" nil
Execute	.target.mkdir "/projects" true
Read	.target.dir "/usr" nil
Execute	.target.mkdir "/usr" true
Read	.target.dir "/pub" nil
Execute	.target.mkdir "/pub" true
Write	.etc.exports [$["allowed":["proj*.local.domain(rw)"], "mountpoint":"/projects"], $["allowed":["*.local.domain(ro)", "@trusted(rw)"], "mountpoint":"/usr"], $["allowed":["(ro,insecure,all_squash)"], "mountpoint":"/pub"]] true
Write	.sysconfig.nfs.NFS4_SUPPORT "no" true
Write	.sysconfig.nfs.NFS_SECURITY_GSS "no" true
Write	.sysconfig.nfs nil true
Read	.init.scripts.exists "rpcbind" true
Read	.init.scripts.exists "rpcbind" true
Read	.init.scripts.exists "rpcbind" true
Read	.init.scripts.exists "rpcbind" true
Read	.init.scripts.runlevel "rpcbind" $["nfslock":$["start":["3", "5"], "stop":["3", "5"]], "nfsserver":$["start":["3", "5"], "stop":["3", "5"]], "rpcbind":$["start":["3", "5"], "stop":["3", "5"]]]
Read	.init.scripts.comment "rpcbind" $["nfslock":$[], "nfsserver":$[], "rpcbind":$[]]
Read	.init.scripts.exists "nfslock" true
Read	.init.scripts.exists "nfslock" true
Read	.init.scripts.exists "nfslock" true
Read	.init.scripts.runlevel "nfslock" $["nfslock":$["start":["3", "5"], "stop":["3", "5"]], "nfsserver":$["start":["3", "5"], "stop":["3", "5"]], "rpcbind":$["start":["3", "5"], "stop":["3", "5"]]]
Read	.init.scripts.comment "nfslock" $["nfslock":$[], "nfsserver":$[], "rpcbind":$[]]
Read	.init.scripts.exists "nfsserver" true
Read	.init.scripts.exists "nfsserver" true
Read	.init.scripts.exists "nfsserver" true
Read	.init.scripts.runlevel "nfsserver" $["nfslock":$["start":["3", "5"], "stop":["3", "5"]], "nfsserver":$["start":["3", "5"], "stop":["3", "5"]], "rpcbind":$["start":["3", "5"], "stop":["3", "5"]]]
Read	.init.scripts.comment "nfsserver" $["nfslock":$[], "nfsserver":$[], "rpcbind":$[]]
Read	.init.scripts.exists "idmapd" true
Execute	.target.bash "/etc/init.d/idmapd status" $["TERM":"raw"] 0
Read	.init.scripts.exists "svcgssd" true
Execute	.target.bash "/etc/init.d/svcgssd status" $["TERM":"raw"] 0
Read	.init.scripts.exists "svcgssd" true
Execute	.target.bash "/etc/init.d/svcgssd stop" $["TERM":"raw"] 0
Read	.init.scripts.exists "rpcbind" true
Execute	.target.bash "/etc/init.d/rpcbind status" $["TERM":"raw"] 0
Read	.init.scripts.exists "nfsserver" true
Execute	.target.bash "/etc/init.d/nfsserver stop" $["TERM":"raw"] 0
Read	.init.scripts.exists "nfslock" true
Execute	.target.bash "/etc/init.d/nfslock restart" $["TERM":"raw"] 0
Read	.init.scripts.exists "nfsserver" true
Execute	.target.bash "/etc/init.d/nfsserver start" $["TERM":"raw"] 0
Read	.init.scripts.exists "nfsserver" true
Execute	.target.bash "/etc/init.d/nfsserver status" $["TERM":"raw"] 0
Read	.init.scripts.exists "SuSEfirewall2_init" true
Execute	.target.bash "/etc/init.d/SuSEfirewall2_init status" $["TERM":"raw"] 0
Return	true
Dump	Write KO
Read	.target.dir "/projects" nil
Execute	.target.mkdir "/projects" true
Read	.target.dir "/usr" nil
Execute	.target.mkdir "/usr" true
Read	.target.dir "/pub" nil
Execute	.target.mkdir "/pub" true
Write	.etc.exports [$["allowed":["proj*.local.domain(rw)"], "mountpoint":"/projects"], $["allowed":["*.local.domain(ro)", "@trusted(rw)"], "mountpoint":"/usr"], $["allowed":["(ro,insecure,all_squash)"], "mountpoint":"/pub"]] false
Log	Unable to write to /etc/exports.
Return	false
